File: frontend/src/features/dsl/api/dslApi.ts
/*
 * Copyright 2026 agwlvssainokuni
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
//
// DSL の管理の API（契約 C6 の 10 本）を呼ぶ関数（frontend-components.md の 1.2、BR2.1〜BR2.5・BR2.8）。
// すべて既存の ApiClient を通し、アクセストークンは ApiClient の中だけで扱う（NFR3.10）。
// 失敗は ApiClient の ApiError（kind・status・code、本文が Problem Details なら problem）として投げる。
import { apiDownload, apiRequest } from '../../../shared/api-client/apiClient'
import { networkError, type ApiError } from '../../../shared/api-client/apiError'
import type {
  DslErrorItem,
  DslErrorReport,
  DslFile,
  DslStatus,
  HistoryEntry,
  Preview,
  SubmitSource,
} from './types'

/** DSL の管理の API の根 */
export const DSL_API_ROOT = '/api/admin/dsl'

/** 各 API のパス */
export const DSL_API_PATHS = {
  status: `${DSL_API_ROOT}/status`,
  preview: `${DSL_API_ROOT}/preview`,
  generate: `${DSL_API_ROOT}/preview/generate`,
  previewDownload: `${DSL_API_ROOT}/preview/download`,
  apply: `${DSL_API_ROOT}/apply`,
  appliedDownload: `${DSL_API_ROOT}/applied/download`,
  history: `${DSL_API_ROOT}/history`,
} as const

/** 履歴の版をプレビューに戻す API のパス */
export function restorePath(revisionId: string): string {
  return `${DSL_API_PATHS.history}/${encodeURIComponent(revisionId)}/restore`
}

/** 投入の本文の形 */
export const YAML_CONTENT_TYPE = 'application/yaml'

/** ファイル名が読めないときのダウンロードのファイル名（BR2.4） */
export const DEFAULT_DOWNLOAD_FILE_NAME = 'dsl.yaml'

/** 本文を JSON として読む。読めなければ通信の失敗として扱う（画面は一般の文言を出す）。 */
async function readJson<T>(response: Response): Promise<T> {
  try {
    return (await response.json()) as T
  } catch {
    const error: ApiError = networkError()
    throw error
  }
}

async function getJson<T>(path: string): Promise<T> {
  return readJson<T>(await apiRequest(path, { method: 'GET' }))
}

async function postJson<T>(path: string, init: RequestInit = {}): Promise<T> {
  return readJson<T>(await apiRequest(path, { ...init, method: 'POST' }))
}

/** 今の状態を読む（GET /status）。 */
export function getStatus(): Promise<DslStatus> {
  return getJson<DslStatus>(DSL_API_PATHS.status)
}

/** プレビューの中身を読む（GET /preview）。無ければ 404 DSL_PREVIEW_NOT_FOUND。 */
export function getPreview(): Promise<Preview> {
  return getJson<Preview>(DSL_API_PATHS.preview)
}

/** スキーマを読み込み、既定の DSL をプレビューに置く（POST /preview/generate）。 */
export function generatePreview(): Promise<Preview> {
  return postJson<Preview>(DSL_API_PATHS.generate)
}

/**
 * DSL を投入する（POST /preview?source=...）。本文は `application/yaml` の文字列で送る（BR2.2）。
 * 文字列の本文は UTF-8 で送られ、長さ（Content-Length）が付く。
 */
export function submitPreview(text: string, source: SubmitSource): Promise<Preview> {
  return postJson<Preview>(`${DSL_API_PATHS.preview}?source=${source}`, {
    headers: { 'Content-Type': YAML_CONTENT_TYPE },
    body: text,
  })
}

/** プレビューを破棄する（DELETE /preview）。無ければ 404 DSL_PREVIEW_NOT_FOUND。 */
export async function discardPreview(): Promise<void> {
  await apiRequest(DSL_API_PATHS.preview, { method: 'DELETE' })
}

/** 表示しているプレビューの識別を付けて適用する（POST /apply、BR2.5）。 */
export function applyPreview(previewId: string): Promise<DslStatus> {
  return postJson<DslStatus>(DSL_API_PATHS.apply, {
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ previewId }),
  })
}

/** 適用の履歴を読む（GET /history、新しい順）。 */
export function getHistory(): Promise<HistoryEntry[]> {
  return getJson<HistoryEntry[]>(DSL_API_PATHS.history)
}

/** 履歴の版をプレビューに戻す（POST /history/{revisionId}/restore）。 */
export function restoreRevision(revisionId: string): Promise<Preview> {
  return postJson<Preview>(restorePath(revisionId))
}

/**
 * `Content-Disposition` からファイル名を取る（`filename*` を優先し、次に `filename`）。
 * 読めなければ {@link DEFAULT_DOWNLOAD_FILE_NAME}。画面でファイル名を組み立てない（BR2.4）。
 */
export function fileNameFromContentDisposition(header: string | null): string {
  if (!header) {
    return DEFAULT_DOWNLOAD_FILE_NAME
  }
  const extended = /filename\*\s*=\s*(?:UTF-8|utf-8)''([^;]+)/.exec(header)
  if (extended) {
    try {
      const decoded = decodeURIComponent(extended[1].trim())
      if (decoded) {
        return decoded
      }
    } catch {
      // 符号化が壊れているときは、次の filename を見る。
    }
  }
  const plain = /filename\s*=\s*(?:"([^"]*)"|([^;\s]+))/.exec(header)
  const name = (plain?.[1] ?? plain?.[2] ?? '').trim()
  return name || DEFAULT_DOWNLOAD_FILE_NAME
}

async function download(path: string): Promise<DslFile> {
  const { blob, contentDisposition } = await apiDownload(path, { method: 'GET' })
  return { blob, fileName: fileNameFromContentDisposition(contentDisposition) }
}

/** プレビュー中の DSL を受け取る（GET /preview/download）。 */
export function downloadPreview(): Promise<DslFile> {
  return download(DSL_API_PATHS.previewDownload)
}

/** 適用中の DSL を受け取る（GET /applied/download）。 */
export function downloadApplied(): Promise<DslFile> {
  return download(DSL_API_PATHS.appliedDownload)
}

function isErrorItem(value: unknown): value is DslErrorItem {
  if (typeof value !== 'object' || value === null) {
    return false
  }
  const item = value as Record<string, unknown>
  const optionalNumber = (v: unknown) => v === undefined || v === null || typeof v === 'number'
  const optionalString = (v: unknown) => v === undefined || v === null || typeof v === 'string'
  return (
    typeof item.kind === 'string' &&
    typeof item.message === 'string' &&
    optionalNumber(item.line) &&
    optionalNumber(item.column) &&
    optionalString(item.path)
  )
}

/**
 * 失敗から誤りの一覧を取り出す（BR2.8）。422 の本文に `total`（1以上の整数）と `errors`（誤りの配列）が
 * あるときだけ返し、形が合わなければ undefined（画面は件数の分からない誤りとして一般の文言を出す）。
 */
export function readErrorReport(error: unknown): DslErrorReport | undefined {
  const apiError = error as ApiError | undefined
  if (apiError?.kind !== 'response' || apiError.problem === undefined) {
    return undefined
  }
  const { total, errors } = apiError.problem
  if (typeof total !== 'number' || !Number.isInteger(total) || total < 1) {
    return undefined
  }
  if (!Array.isArray(errors) || !errors.every(isErrorItem)) {
    return undefined
  }
  const items = errors.map((item) => ({
    kind: item.kind,
    line: item.line ?? null,
    column: item.column ?? null,
    path: item.path ?? null,
    message: item.message,
  }))
  return { total, errors: items }
}

/** 画面が使う API の関数の集まり（テストで差し替える） */
export const dslApi = {
  getStatus,
  getPreview,
  generatePreview,
  submitPreview,
  discardPreview,
  applyPreview,
  getHistory,
  restoreRevision,
  downloadPreview,
  downloadApplied,
}

/** 画面が使う API の関数の集まりの型 */
export type DslApi = typeof dslApi

File: backend/src/test/java/cherry/mastersmith/dslmanage/testsupport/DslApi.java
/*
 * Copyright 2026 agwlvssainokuni
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package cherry.mastersmith.dslmanage.testsupport;

import cherry.mastersmith.common.testsupport.HttpTestClient;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.UncheckedIOException;
import java.net.Socket;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import tools.jackson.core.type.TypeReference;
import tools.jackson.databind.json.JsonMapper;

/** DSL の管理の API を、アクセストークンつきで呼ぶテストの補助（実際の番号で待ち受けるアプリへ送る）。 */
public final class DslApi {

    /** API の根。 */
    public static final String ROOT = "/api/admin/dsl";

    private static final JsonMapper MAPPER = JsonMapper.builder().build();

    private final HttpTestClient client;

    private final String accessToken;

    private final String language;

    private final int port;

    /**
     * 作る。
     *
     * @param port アプリの待ち受けの番号
     * @param accessToken アクセストークン（無ければ null。未認証の要求）
     */
    public DslApi(int port, String accessToken) {
        this(port, accessToken, null);
    }

    private DslApi(int port, String accessToken, String language) {
        this.client = new HttpTestClient(port);
        this.accessToken = accessToken;
        this.language = language;
        this.port = port;
    }

    /**
     * 表示言語（Accept-Language）を変えた補助を返す。
     *
     * @param acceptLanguage Accept-Language の値
     * @return 補助
     */
    public DslApi withLanguage(String acceptLanguage) {
        return new DslApi(port, accessToken, acceptLanguage);
    }

    private HttpRequest.Builder request(String path) {
        HttpRequest.Builder builder = client.request(ROOT + path);
        if (accessToken != null) {
            builder.header("Authorization", "Bearer " + accessToken);
        }
        if (language != null) {
            builder.header("Accept-Language", language);
        }
        return builder;
    }

    /**
     * GET を送る。
     *
     * @param path {@code /api/admin/dsl} からのパス
     * @return 応答
     */
    public HttpResponse<String> get(String path) {
        return client.send(request(path).GET().build());
    }

    /**
     * GET を送り、本文をバイト列で受け取る。
     *
     * @param path {@code /api/admin/dsl} からのパス
     * @return 応答
     */
    public HttpResponse<byte[]> getBytes(String path) {
        try {
            return HttpClient.newHttpClient()
                    .send(request(path).GET().build(), HttpResponse.BodyHandlers.ofByteArray());
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException(e);
        }
    }

    /**
     * DSL を投入する（{@code application/yaml}）。
     *
     * @param body 本文
     * @param source 出どころ（UPLOAD・PASTE）
     * @return 応答
     */
    public HttpResponse<String> submit(byte[] body, String source) {
        return submit(body, source, "application/yaml");
    }

    /**
     * 本文の形を指定して DSL を投入する。
     *
     * @param body 本文
     * @param source 出どころ
     * @param contentType 本文の形
     * @return 応答
     */
    public HttpResponse<String> submit(byte[] body, String source, String contentType) {
        return client.send(request("/preview?source=" + source)
                .header("Content-Type", contentType)
                .POST(HttpRequest.BodyPublishers.ofByteArray(body))
                .build());
    }

    /**
     * 本文を送らずに、{@code Content-Length} のヘッダーだけで投入の要求を送り、応答を受け取る。本文を読まずに断ること（413・401）を
     * 確かめるために使う（アプリが本文を読もうとすれば、送られないため応答が返らず、読み取りの待ちの上限で失敗する）。
     *
     * @param contentLength 名乗る本文の大きさ
     * @param source 出どころ
     * @return 状態コードと本文
     */
    public RawResponse submitHeadersOnly(long contentLength, String source) {
        try (Socket socket = new Socket("localhost", port)) {
            socket.setSoTimeout(15_000);
            StringBuilder head = new StringBuilder()
                    .append("POST ")
                    .append(ROOT)
                    .append("/preview?source=")
                    .append(source)
                    .append(" HTTP/1.1\r\n")
                    .append("Host: localhost:")
                    .append(port)
                    .append("\r\n")
                    .append("Content-Type: application/yaml\r\n")
                    .append("Content-Length: ")
                    .append(contentLength)
                    .append("\r\n")
                    .append("Connection: close\r\n");
            if (accessToken != null) {
                head.append("Authorization: Bearer ").append(accessToken).append("\r\n");
            }
            head.append("\r\n");
            OutputStream out = socket.getOutputStream();
            out.write(head.toString().getBytes(StandardCharsets.US_ASCII));
            out.flush();
            InputStream in = socket.getInputStream();
            int status = Integer.parseInt(readLine(in).split(" ")[1]);
            int length = -1;
            boolean chunked = false;
            String line;
            while (!(line = readLine(in)).isEmpty()) {
                String lower = line.toLowerCase(Locale.ROOT);
                if (lower.startsWith("content-length:")) {
                    length = Integer.parseInt(
                            line.substring(line.indexOf(':') + 1).strip());
                } else if (lower.startsWith("transfer-encoding:") && lower.contains("chunked")) {
                    chunked = true;
                }
            }
            ByteArrayOutputStream body = new ByteArrayOutputStream();
            if (chunked) {
                int size;
                while ((size = Integer.parseInt(readLine(in).split(";")[0].strip(), 16)) > 0) {
                    body.write(in.readNBytes(size));
                    readLine(in);
                }
            } else if (length > 0) {
                body.write(in.readNBytes(length));
            }
            return new RawResponse(status, body.toString(StandardCharsets.UTF_8));
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    /** CRLF で終わる1行を読む（行の終わりの記号を除く）。 */
    private static String readLine(InputStream in) throws IOException {
        ByteArrayOutputStream line = new ByteArrayOutputStream();
        int b;
        while ((b = in.read()) != -1 && b != '\n') {
            if (b != '\r') {
                line.write(b);
            }
        }
        return line.toString(StandardCharsets.UTF_8);
    }

    /**
     * 本文を送らない要求の応答。
     *
     * @param status 状態コード
     * @param body 本文
     */
    public record RawResponse(int status, String body) {}

    /**
     * スキーマを読み込む。
     *
     * @return 応答
     */
    public HttpResponse<String> generate() {
        return client.send(request("/preview/generate")
                .POST(HttpRequest.BodyPublishers.noBody())
                .build());
    }

    /**
     * プレビューを破棄する。
     *
     * @return 応答
     */
    public HttpResponse<String> discard() {
        return client.send(request("/preview").DELETE().build());
    }

    /**
     * 適用する。
     *
     * @param previewId 見たプレビューの識別
     * @return 応答
     */
    public HttpResponse<String> apply(String previewId) {
        return post("/apply", MAPPER.writeValueAsString(Map.of("previewId", previewId)));
    }

    /**
     * 履歴の版をプレビューに戻す（本文なし）。
     *
     * @param revisionId 戻す版の識別
     * @return 応答
     */
    public HttpResponse<String> restore(String revisionId) {
        return client.send(request("/history/" + revisionId + "/restore")
                .POST(HttpRequest.BodyPublishers.noBody())
                .build());
    }

    /**
     * JSON の本文で POST を送る。
     *
     * @param path {@code /api/admin/dsl} からのパス
     * @param json 本文
     * @return 応答
     */
    public HttpResponse<String> post(String path, String json) {
        return client.send(request(path)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build());
    }

    /**
     * 投入して、置かれたプレビューの識別を返す（201 でなければ失敗させる）。
     *
     * @param body 本文
     * @return previewId
     */
    public String submitOk(byte[] body) {
        HttpResponse<String> response = submit(body, "PASTE");
        if (response.statusCode() != 201) {
            throw new IllegalStateException("投入に失敗しました: " + response.statusCode() + " " + response.body());
        }
        return (String) json(response).get("previewId");
    }

    /**
     * JSON の対応表にする。
     *
     * @param response 応答
     * @return 対応表
     */
    public static Map<String, Object> json(HttpResponse<String> response) {
        return HttpTestClient.json(response);
    }

    /**
     * JSON の並びにする。
     *
     * @param response 応答
     * @return 並び
     */
    public static List<Map<String, Object>> jsonList(HttpResponse<String> response) {
        return MAPPER.readValue(response.body(), new TypeReference<List<Map<String, Object>>>() {});
    }
}
